c2corg_ui/static/js/searchdirective.js

```diff
@@ -0,0 +1,184 @@
+ * @param {angular.$compile} $compile Angular compile service.
+ * @param {string} apiUrl Base URL of the API.
+ * @param {angularGettext.Catalog} gettextCatalog Gettext catalog.
+ * @param {Array.<string>} cultures All cultures.
+ * @ngInject
+ */
+app.SearchController = function(
+    $rootScope, $compile, apiUrl, gettextCatalog, cultures) {
+
+  /**
+   * @type {string}
+   * @private
+   */
+  this.apiUrl_ = apiUrl;
+
+  /**
+   * @type {Array.<string>}
+   * @private
+   */
+  this.cultures_ = cultures;
+
+  /**
+   * @type {angularGettext.Catalog}
+   * @private
+   */
+  this.gettextCatalog_ = gettextCatalog;
+
+  /**
+   * @type {TypeaheadOptions}
+   * @export
+   */
+  this.options = /** @type {TypeaheadOptions} */ ({
+    highlight: true,
+    hint: true,
+    minLength: 2
+  });
+
+  /**
+   * @type {Array.<TypeaheadDataset>}
+   * @export
+   */
+  this.datasets = [
+    this.createDataset_('waypoints', 'Waypoints'),
+    this.createDataset_('routes', 'Routes')
+  ];
+
+  /**
+   * @type {ngeox.SearchDirectiveListeners}
+   * @export
+   */
+  this.listeners = /** @type {ngeox.SearchDirectiveListeners} */ ({
+    select: angular.bind(this, app.SearchController.select_)
+  });
+};
+
+
+/**
+ * @param {string} type The document type.
+ * @param {string} title The category title.
+ * @return {TypeaheadDataset} A data set.
+ * @private
+ */
+app.SearchController.prototype.createDataset_ = function(type, title) {
+  var bloodhoundEngine = this.createAndInitBloodhound_(type);
+  return /** @type {TypeaheadDataset} */({
+    source: bloodhoundEngine.ttAdapter(),
+    display: function(suggestion) {
+      return suggestion['label'];
+    },
+    limit: 20,
+    templates: {
+      header: angular.bind(this, function() {
+        return '<div class="header">' +
+            this.gettextCatalog_.getString(title) + '</div>';
+      }),
+      suggestion: function(suggestion) {
+        return '<p>' + suggestion['label'] + '</p>';
+      }
+      // Note: The templates for `notFound` and `pending` are not set, because
+      // they would be shown for each dataset.
+    }
+  });
+};
+
+
+/**
+ * @param {string} type The document type.
+ * @return {Bloodhound} The bloodhound engine.
+ * @private
+ */
+app.SearchController.prototype.createAndInitBloodhound_ = function(type) {
+  var url = this.apiUrl_ + '/search?q=%QUERY';
+  var bloodhound = new Bloodhound(/** @type {BloodhoundOptions} */({
+    limit: 10,
+    queryTokenizer: Bloodhound.tokenizers.whitespace,
+    datumTokenizer: Bloodhound.tokenizers.obj.whitespace('label'),
+    remote: {
+      url: url,
+      wildcard: '%QUERY',
+      rateLimitWait: 50,
+      filter: angular.bind(this, function(resp) {
+        var documents = resp[type]['documents'];
+        var currentCulture = this.gettextCatalog_.currentLanguage;
+        var cultures = this.cultures_;
+        return documents.map(function(document) {
+          var locale = app.utils.getBestLocale(
```

Comment:
We should have a typedef for this kind of document response and not use brackets.
Typedefs can be added to c2corg_ui/externs/appx.js